src/commands/Find.thy

```diff
@@ -0,0 +1,75 @@
+theory Find
+  imports Main
+begin
+
+section \<open>Find theorems with @{command find_theorems}\<close>
+text \<open>
+  Theorems can be searched by name where @{verbatim *} wildcards are allowed.
+  One of the theorems found with the query below is @{thm gauss_sum_nat}.
+\<close>
+find_theorems name: "gau*"
+
+text \<open>
+  You can also search theorems by specifying an arbitrary number of patterns.
+  Patterns are allowed to contain wildcards (@{verbatim _}) constants, schematic variables, and types.
+  It is also possible to negate a pattern by prefixing it with a @{verbatim -}.
+  If a negated pattern occurs in a theorem, it will not be considered by the search.
+\<close>
+find_theorems "finite (?A::'a rel)"
+find_theorems "finite _ \<Longrightarrow> _" "insert _ _" -"card _"
+find_theorems "finite ?A \<Longrightarrow> finite ?B \<Longrightarrow> _ (?A \<union> ?B)"
+
+text \<open>
+  In addition to patterns, we can take the goal state into account to search for theorems.
+  The example below uses intro to search for introduction rules that solve the current goal.
+  For the given goal, @{command find_theorems} finds the lemma @{thm bij_betw_byWitness} which we instantiate accordingly.
```

Comment:
Typically, Isabelle users try to stay within the 100 characters line limit indicated by the blue line in jEdit.